[Typechecker] Check for generic signature when checking for overrides

theblixguy opened the pull request:
We currently do not check for generic signatures when performing override checking. This PR adds a check for generic signature to `OverrideMatcher::checkOverride()`.

Resolves [SR-4206](https://bugs.swift.org/browse/SR-4206), [SR-4986](https://bugs.swift.org/browse/SR-4986), [SR-7573](https://bugs.swift.org/browse/SR-7573), [SR-10076](https://bugs.swift.org/browse/SR-10076), [SR-10198](https://bugs.swift.org/browse/SR-10198) & [SR-10603](https://bugs.swift.org/browse/SR-10603)
Resolves rdar://problem/23626260
Resolves rdar://problem/32378463
Resolves rdar://problem/39868032
Resolves rdar://problem/49339618

theblixguy committed:
[Typechecker] Check for generic signature requirements when pechecking for overrides

theblixguy committed:
[Test] Add tests

theblixguy commented:
cc @jrose-apple @slavapestov

theblixguy committed:
[Test] Remove dead test code

slavapestov commented on test/attr/attr_override.swift:
```diff
@@ -504,3 +504,32 @@ class Index<F, T> {
+class SR_4206_C5 {
+  func test<E: SR_4206_C3>(_: E) {} // expected-note {{overridden declaration is here}}
+}
+class SR_4206_C6<T>: SR_4206_C5 {
```
Can you also test the other direction, where the concrete class inherits from a generic class?

theblixguy replied:
Done!

slavapestov reviewed (approved):
Nice!

theblixguy committed:
[Test] Adds a test case where a concrete class inherits from a generic class

theblixguy committed:
[Typechecker] Check if the generic signature is null before accessing it

theblixguy commented:
Hmm building the stdlib is producing the new diagnostics. Here's an example of code that triggers it:

```swift
protocol CodingKey {}

protocol Container {
  associatedtype Key: CodingKey
}

class Base<Key: CodingKey> {
  func foo(forKey key: Key) throws {}
}

class Derived<Concrete: Container> : Base<Concrete.Key> {
  typealias Key = Concrete.Key
  override func foo(forKey key: Key) throws {}
}
```

`Concrete.Key` conforms to `CodingKey` so it should satisfy the requirement `Key: CodingKey` on `Base`.

slavapestov commented:
Oh, I see. I guess it's not sufficient to ask if the base signature has any requirements not satisfied by the derived signature. You need to apply the superclass type context substitutions (`derivedClass->getSuperclass()->getContextSubstitutionMap(M, baseClass)`) to each base requirement before checking it against the derived signature.

Adding a SubstitutionMap argument to `requirementsNotSatisfiedBy()` is your best bet. Do you also mind adding a reduced test case for the stdlib failure to the test suite, so that in the future we can catch a regression here without rebuilding the stdlib?

theblixguy committed:
[Typechecker] Pass a substitution map to requirementsNotSatisfiedBy()

theblixguy committed:
[Test] Adds a new test case that mimics a Codable class

This is basically  a sort of regression test that ensures we apply proper context substitutions before checking the generic signatures of the functions

theblixguy commented:
Thanks @slavapestov, that seems to work. I have added the above example to the test.

theblixguy committed:
[Test] Fix a typo - use expected-note and not expected-error

jrose-apple commented:
Will this break any working code? (That's a little different from "code that compiles today".)

@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test

slavapestov commented on lib/AST/GenericSignature.cpp:
```diff
@@ -603,7 +604,14 @@ SmallVector<Requirement, 4> GenericSignature::requirementsNotSatisfiedBy(
+    auto reqToCheck = req;
+    if (subMap) {
+      auto subsReq = req.subst(subMap);
+      if (subsReq.hasValue()) {
```
If the substitution failed... I'm not sure what to do. Using the unsubstituted requirement is clearly wrong. Maybe just return assert in this case, and we'll rethink it if/when we hit it?

theblixguy replied:
Sure, I have added an assert!

slavapestov reviewed (approved):
Thank you. One more optional suggestion but you can address it in a follow up PR if you don't want to do it now.

slavapestov commented:
@theblixguy It looks like you're checking that the base signature does not have any requirements not satisfied by the derived method signature. That is, we don't allow the derived method to have a more general generic signature.

We actually want to ban the other direction too, for example this crashes:

```
class A {
  func f<T : P>(_: T) {}
}

class B : A {
  override func f<T : P & Q>(_ t: T) {
    t.doit()
  }
}

protocol P {}
protocol Q {
  func doit()
}

let b: A = B()

struct R : P, Q {
  func doit() {
    print("hi")
  }
}

b.f(R())
```

I think we want the generic requirements to match exactly -- if the base is more general than derived, you can't call a base method on a derived value in a sound way. On the other hand if the base is less general than derived, then super.foo() calls can't be made from the derived method.

@jrose-apple What do you think? And yeah, it's possible that this makes us now reject formerly-working code, but that's really iffy. We rely on the ABI of a derived method matching a base method exactly (if not, we emit a vtable thunk, but the ABI check does not look at generic requirements either!) 

It's _possible_ that things line up somehow even when the requirements don't match (eg, superclass constraints and @objc protocol constraints don't introduce witness table parameters) but... ugh.

theblixguy committed:
[Typechecker] Make sure superclass isn't null before accessing it

jrose-apple commented:
Liskov substitutability says we need Derived to be at least as general as Base. The `super.foo()` thing is interesting because we allow other types of covariance and restrict the `super` call when you try to make it, but we don't have a way to refine generic parameters that way—and as Slava points out, we'd have to go further to make sure the vtable builder added a new slot for the new signature. Let's just stick to exact matches for now.

jrose-apple commented:
I suppose we should run this by the core team as an acceptable source-breaking change (I agree that it's worth it), and possibly add it to CHANGELOG as well.

slavapestov commented:
@jrose-apple Right, I didn't realize that the super.foo() thing would come up if you, eg, made a parameter more optional as well, which we _do_ allow. So I guess the right answer is for the vtable check to add a new entry if the base requirements are not satisfied by the derived signature. Only derived requirements not being satisfied by the base signature has to be rejected for soundness reasons.

@theblixguy This is totally optional, but if you wanted to try implementing that, take a look at `requiresNewVTableEntry()` in lib/AST/Decl.cpp. We strip off the generic signatures before calling TypeBase::matches() on the base and derived type, so you'll need to check for them separately before or after doing that. Then to test it, you'll need to write some SILGen tests (and execution tests too preferably). But like I said, it's totally reasonable to land this fix first (but note that you'll need to change the polarity of your 'requirements satisfied' check, since the case you want to ban is when the derived signature has requirements not satisfied by the base, not the other way around).

theblixguy committed:
[Typechecker] Check for exact generic signature

theblixguy commented:
Sounds good to me, I have changed it so we now ensure there's an exact match.

slavapestov commented:
@theblixguy My apologies for the confusion but "exact match" means you still have to areRequirementsSatisfiedBy() with the substitution map, but you're checking both signatures.

slavapestov commented:
(Try running the tests with your current version, they should hopefully fail)

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -943,6 +944,22 @@ bool OverrideMatcher::checkOverride(ValueDecl *baseDecl,
+    auto baseGenericSig = baseGenericCtx->getGenericSignature();
+    auto declGenericSig = declGenericCtx->getGenericSignature();
+    
+    if (baseGenericSig != declGenericSig) {
```
baseGenericSig should not have any requirements not satisfied by declGenericSig, or vice versa.

Now it's only possible to build a substitution map in one direction (turning base class requirements into derived class requirements). So requirementsNotSatisfiedBy will need to take _two_ substitution maps; the first one is applied to the base requirement, the second one is applied to the derived requirement.

Your two calls both pass one empty and one non-empty substitution map, flipped around.

theblixguy replied:
Can I just call `baseGenericSig->requirementsNotSatisfiedBy(declGenericSig, subMap)` & declGenericSig->requirementsNotSatisfiedBy(baseGenericSig, subMap)` instead? For `subMap`, I can just flip it around by using `baseClass` instead of `derivedClass`.

theblixguy replied:
Hmm that seems to work on tests, but fails when building the stdlib.

theblixguy replied:
How can two maps be used inside `requirementsNotSatisfiedBy()`? Because it uses `otherSig` and calls `isRequirementSatisfied` on it and the `req` passed is from the generic signature you call `requirementsNotSatisfiedBy()` on.

theblixguy replied:
@slavapestov Could you clarify what you meant when you said "So requirementsNotSatisfiedBy will need to take two substitution maps"? When I call `requirementsNotSatisfiedBy()`, it's easy to apply a substitution to the signature I pass to it. To apply a substitution to the requirements of the other signature, I would need to extract its requirements, substitute them and then re-create a new signature (maybe skip adding the requirement if it can't be substituted?). 

Is that what you meant? Thank you!

theblixguy committed:
[Typechecker] Check that either base requirements are not satisfied or derived requirements are not satisfied

theblixguy committed:
[GenericSignature] Fix indentation

slavapestov commented:
Sorry for neglecting this. I will take a look as this in more detail as soon as I can.

In the mean time, here's a brain dump. I think you're right that building a new signature is your best bet.

So consider you have a class hierarchy like

```
class Base<T : IteratorProtocol, U> {
  func foo<V : IteratorProtocol, V.Element == T.Element>(_: U) {}
}

class Derived<T : Collection> : Base<T.Iterator, Int> {
  // note: T.Element should work as well
  override func foo<V : IteratorProtocol, U`V.Element == T.Iterator.Element>(_: U) {}
}
```

The superclass substitutions are:
```
T_0_0 := T_0_0.Element
T_0_1 := Int
T_0_0 : Iterator := conformance of T_0_0.Iterator to IteratorProtocol, derived from conformance of T_0_0 to Collection
```

Instead of applying substitutions to each requirement of the base signature before testing if the derived signature satisfies it, which only works in one direction, we build a whole new signature from the base signature.

A generic signature has parameters and requirements. So you have a generic signature for the base method. Zero or more generic parameters at the start are from the base generic type; you drop them. In their place, you add the generic parameters from the derived generic type. Then, you take any remaining generic arguments from the base method signature, and you "shift" them by adjusting their depth as necessary. To see why you have to do this, consider these cases:

```
class Generic<T> { func foo<U>(_: ...) {} } // depth is 1
class Concrete : Generic<T> { override func foo<U>(_: ...) {} } // depth is 0

class Outer<T> { class Inner<U> : Generic<(T, U)> { ... } }
```

The requirements then get mapped by applying the substitution map to each one. However, note that the substitution map only maps the base type's generic parameters. You'll in fact need to build a _new_ substitution map that handles all of the base method's generic parameters -- mapping the type's parameters to the corresponding replacements from the superclass substitutions, and mapping the method's parameters by 'shifting' the depth as needed while preserving the index. I think getOverrideSubstitutions() implements this already, so perhaps just grabbing the override substitutions (they might be computed in nearby code already) and applying them to the requirements will do the trick. Also I would not expect this substitution to fail, except on invalid code. It's fine to say the override doesn't match in this case.

So this is a bit more involved than I imagined. But now you will have a generic signature that combines the derived type's parameters with the base method's parameters and requirements.

It should then be possible to pass this generic signature together with the derived method signature to requirementsNotSatisfiedBy(), and critically, test both directions of the relation.

theblixguy commented:
Thank you @slavapestov. Do you mean something like this?

```c++
static bool areRequirementsSatisfied(ValueDecl *base, ValueDecl *derived) {

  auto baseGenericCtx = base->getAsGenericContext();
  auto derivedGenericCtx = derived->getAsGenericContext();

  if (baseGenericCtx && derivedGenericCtx) {
    auto baseGenericSig = baseGenericCtx->getGenericSignature();
    auto derivedGenericSig = derivedGenericCtx->getGenericSignature();

    auto baseClass = baseGenericCtx->getInnermostTypeContext();
    auto derivedClass = derivedGenericCtx->getInnermostTypeContext();

    if (baseClass && derivedClass) {
      if (isa<ClassDecl>(baseClass) && isa<ClassDecl>(derivedClass)) {
        auto derivedSuperclass =
            derivedClass->getSelfClassDecl()->getSuperclass();
        auto derivedSubMap = derivedSuperclass
                                 ? derivedSuperclass->getContextSubstitutionMap(
                                       base->getModuleContext(), baseClass)
                                 : SubstitutionMap();

        auto subMap = SubstitutionMap::getOverrideSubstitutions(base, derived,
                                                                derivedSubMap);
        auto baseReqsSatisfied = baseGenericSig->requirementsNotSatisfiedBy(derivedGenericSig, subMap).empty();
        auto derivedReqsSatisfied = derivedGenericSig->requirementsNotSatisfiedBy(baseGenericSig, subMap).empty();
        return baseReqsSatisfied && derivedGenericSig;
      }
    }
  }

  return true;
}
```

slavapestov commented:
@theblixguy I don't think that's right because you're using the same substitution map with baseGenericSig and derivedGenericSig, which is an error.

Take a look at how configureGenericDesignatedInitOverride() builds the generic signature for the inherited designated initializer. I think this is what you want to do as well.

theblixguy commented:
@slavapestov In your previous comment, you mentioned

> I think getOverrideSubstitutions() implements this already, so perhaps just grabbing the override substitutions (they might be computed in nearby code already) and applying them to the requirements will do the trick.

which is what I was doing above - except I am also using the same subMap in both directions. Should I just use the subMap when calling `baseGenericSig->requirementsNotSatisfiedBy` but not the other way round? 

Also, it seems like `getOverrideSubstitutions ` is already doing what `configureGenericDesignatedInitOverride` does.

slavapestov commented:
@theblixguy We need to build a new generic signature using the same logic as `configureGenericDesignatedInitOverride()`. The signature has to combine the requirements of the derived method's generic type with the base method's generic requirements. Then we check if this signature has any requirements not satisfied by the derived method's signature, and vice versa.

We can avoid code duplication by factoring out the code from `configureGenericDesignatedInitOverride()` into a new method.

theblixguy committed:
[Typechecker] Attempt to extract the logic from configureGenericDesignatedInitOverride and tweak it to work with generic methods

theblixguy commented:
Hmm thanks you @slavapestov :) I am still a bit confused because the logic in `configureGenericDesignatedInitOverride` method is based on things like the `superclassCtor` and `superclassTy`, so I have been trying to adjust that to accept the `baseMethod` and `baseMethodTy` instead. Could you take a look at commit #bd2cfad?

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -926,6 +929,84 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
   }
 }
 
+static GenericSignature *getNewGenericSignature(ValueDecl *base,
```
This looks good. Can you try using this new function in CodeSynthesis.cpp as well?

theblixguy replied:
@slavapestov So sorry for the late reply (wwdc had me distracted)!

I can, where do you want me to use this method in CodeSynthesis?

slavapestov replied:
I was thinking it would be used by configureGenericDesignatedInitOverride(). Does that make sense? It looks like it's almost the same computation.

theblixguy replied:
@slavapestov Yeah it would be great to get this PR merged first and then I can clean up CodeSynthesis and do the vtable check thing in separate PRs :)

slavapestov commented:
configureGenericDesignatedInitOverride() is entirely analogous except that superclassCtor is the base class constructor instead of the overridden method, and superclassTy is the base class type  whereas baseMethodTy is the (entire) type of the base method.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

theblixguy committed:
[Typechecker] Fix a few crashes in getNewGenericSignature

theblixguy commented:
@slavapestov the builds failed because of a few crashes in `getNewGenericSignature()`. I have fixed the crashes now, but seems like the tests are failing because the new diagnostic isn't being emitted. One of the checks I had to add to fix the crash was:

> if (!derivedClass->getInterfaceType()->is<LValueType>()) { return nullptr; }

otherwise it was hitting an assert in `TypeSubstitutionMap TypeBase::getContextSubstitutions`. Seems like the test code is hitting that branch and exiting early, hence why the diagnostics aren't being emitted.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -937,11 +937,23 @@ static GenericSignature *getNewGenericSignature(ValueDecl *base,
       derivedGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
 
+  if (!derivedClass) {
+    return nullptr;
```
What's an example where this happens? This code should only be run for class members.

theblixguy replied:
@slavapestov I don’t have a concrete example, but these crashes were happening due to code in the standard library (collections iirc).

slavapestov replied:
You should try to reduce a test case to see what's going on. Maybe the declaration in question is a protocol member, in which case we should skip most of these checks.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -937,11 +937,23 @@ static GenericSignature *getNewGenericSignature(ValueDecl *base,
+    return nullptr;
+  }
+
+  if (!derivedClass->getInterfaceType()->is<LValueType>()) {
```
This is always going to be true, because the type of a ClassDecl is never an LValueType.

theblixguy replied:
@slavapestov Right, so does that mean we should remove the assert from getContextSubstitutions()? Otherwise it’s going to hit the assert there.

slavapestov replied:
The assert in getContextSubstitutionMap() is correct. However getInterfaceType() on a ClassDecl should never return an LValueType. Can you print the decl and its type to verify what's going on?

slavapestov replied:
I see the problem. When you compute 'subMap' below, you should use `derivedClass->getSuperclass()` and not `derivedClass->getInterfaceType()`.

The assert that is firing is that you're passing a MetatypeType to getContextSubstitutions(), not an LValueType.

theblixguy replied:
@slavapestov Ah ok. But then the code causes a crash while doing `for (auto reqt : baseClassSig->getRequirements()) { ... }` (it's calling `getRequirementsBuffer` which causes the crash). I think it makes sense for `baseClassSig` to be null if the class itself doesn't have any generic parameters:

```swift
protocol SR_4206_P0 {}

protocol SR_4206_P1: SR_4206_P0 {}
protocol SR_4206_P2: SR_4206_P0 {}

class SR_4206_C1 {
  func foo<T: SR_4206_P1>(param: T) {} // expected-note {{overridden declaration is here}}
}

class SR_4206_C2: SR_4206_C1 {
  override func foo<T: SR_4206_P2>(param: T) {} // expected-error {{cannot override method 'foo' with a different generic signature}}
}
```

Should we be looping over the method's generic requirements instead in this case?

slavapestov commented on test/attr/attr_override.swift:
```diff
@@ -505,3 +505,53 @@ class Index<F, T> {
+
+class SR_4206_Derived<C: SR_4206_Container> : SR_4206_Base<C.Key> {
+  typealias Key = C.Key
+  override func foo(forKey key: Key) throws {} // Ok
```
Can you add some more tests? We should cover all combinations of:

- Base class is generic/concrete
- Subclass is generic/concrete
- Method is generic/concrete
- Override is valid/invalid

theblixguy replied:
Sure, I'll add more tests once I can get the whole thing to build and pass existing tests 😅

slavapestov replied:
Sure, but keep in mind that adding more precise tests might make it more apparent what the existing problems in the code are, too

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,96 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+      depth = genericSig->getGenericParams().back()->getDepth() + 1;
+
+    for (auto *param : genericParams->getParams()) {
+      auto *newParam = new (ctx) GenericTypeParamDecl(
```
You don't need to create new GenericTypeParamDecls. In fact it appears that newParams is not used below.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,96 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+  }
+
+  auto baseClass =
+      baseGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
```
Since base and derived are always members of a class, I would prefer `baseClass = base->getDeclContext()->getSelfClassDecl()`

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,96 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+      return ProtocolConformanceRef(proto);
+    };
+
+    for (auto reqt : baseClassSig->getRequirements())
```
This should be looping over the base method's requirements, not the base class requirements, so eg baseGenericCtx->getGenericSignature()->getRequirements()

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,96 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+      derivedClass->getModuleContext(), baseClass);
+
+  auto *genericParams = baseGenericCtx->getGenericParams();
+  if (genericParams) {
```
I think the condition you care about is that the base method is in a generic context, not that the base class has its own generic parameters. Eg,

```
class Outer<T> {
  class Inner {
    func method()
  }
}
```

theblixguy replied:
@slavapestov Do you mean something like `base->getDeclContext()->isGenericContext()`? Although that will fail for all base classes that aren't generic (but the method is) and thus this function would bail out and return nullptr.

```swift
protocol SR_4206_P0 {}

protocol SR_4206_P1: SR_4206_P0 {}
protocol SR_4206_P2: SR_4206_P0 {}

/*class SR_4206_C1 {
  func foo<T: SR_4206_P1>(param: T) {} // expected-note {{overridden declaration is here}}
}

class SR_4206_C2: SR_4206_C1 {
  override func foo<T: SR_4206_P2>(param: T) {} // expected-error {{cannot override method 'foo' with a different generic signature}}
}*/
```

The method isn't in a generic context here for example.

slavapestov replied:
No, you want `baseGenericCtx->getGenericSignature() != nullptr`, precisely because the method might have its own generic signature.

slavapestov replied:
The generic signature being built here should consist of all the generic parameters of the base class, plus those from the method that are not also from the derived class. So the list of parameters built should never be totally empty.

theblixguy committed:
[Typechecker] Apply review comments

theblixguy committed:
[Typechecker] No need to assign to subMap as its not used afterwards

theblixguy committed:
[Typechecker] Call baseGenericCtx->isGeneric() before calling getGenericParams() otherwise it can lead to a crash

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -963,7 +963,8 @@ static GenericSignature *getNewGenericSignature(ValueDecl *base,
     unsigned superclassDepth = 0;
-    if (baseClassSig) {
+    
+    if (baseGenericCtx->isGeneric()) {
```
I think the old condition was correct. However, `baseGenericCtx->getGenericParams()` should be `baseClassSig->getParams()`.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,78 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+  auto subMap = derivedClass->getSuperclass()->getContextSubstitutionMap(
+      derivedClass->getModuleContext(), baseClass);
+
+  if (baseGenericCtx) {
```
You already ruled out baseGenericCtx being null on line 939

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -945,6 +1020,27 @@ bool OverrideMatcher::checkOverride(ValueDecl *baseDecl,
+      auto satOne = derivedSig->requirementsNotSatisfiedBy(newSig).empty();
+      auto satTwo = newSig->requirementsNotSatisfiedBy(derivedSig).empty();
+
+      if (!satOne && !satTwo) {
```
Shouldn't this be || and not &&?

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2337,6 +2337,9 @@ WARNING(override_swift3_objc_inference,none,
         (DescriptiveDeclKind, DeclName, Identifier))
+ERROR(override_method_different_generic_sig,none,
+      "cannot override method %0 with a different generic signature",
+      (DeclBaseName))
```
This diagnostic would be better if it actually contained the generic signatures printed out as StringRefs. Eg, "override method''s generic signature %0 is not compatible with base method generic signature %1"

theblixguy committed:
Apply review feedback and update diagnostic

theblixguy commented:
Thank you @slavapestov! Seems like we're really close to get this working (got some of the tests to correctly trigger the new diagnostic). One of the tests still causes a crash:

```
class SR_4206_C1 {
  func foo<T: SR_4206_P1>(param: T) {}
}

class SR_4206_C2: SR_4206_C1 {
  override func foo<T: SR_4206_P2>(param: T) {}
}
```

It causes an assertion to be triggered: `Assertion failed: (!params.empty()), function get, file /Users/suyashsrijan/Documents/swift-src/swift/lib/AST/ASTContext.cpp, line 3841.`. This happens when `computeGenericSignature()` calls `GenericSignature::get`. Is there anything that I am doing wrong?

slavapestov commented:
It sounds like you're not adding the generic parameters of the method itself to the new generic signature that is being built?

theblixguy committed:
[Typechecker] Add derived method's generic parameters to GSB

theblixguy committed:
[Typechecker] Fix a crash when the superclass type is null

theblixguy committed:
[Test] Add some more tests

theblixguy commented:
@slavapestov I think that was it! Can you take another look? I have added some more tests as well.

theblixguy committed:
[Test] Fix test class names

slavapestov commented:
I think this looks good! I'll merge it if the tests pass :)

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

theblixguy committed:
[Typechecker] Return nullptr if base class is not present

theblixguy commented:
> `swift/stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift:145:30: error: overriden method 'dispatchTester' has generic signature <C where C : Collection> which is incompatible with base method's generic signature <S where S : Sequence>`

Hmm I don't think this is expected as `Collection` refines `Sequence`? Small reproducer:

```swift
protocol A {}
protocol B: A {}

class Foo {
  func someMethod<T: A>(arg: T) {}
}

class Bar: Foo {
  override func someMethod<T: B>(arg: T) {}
}
```

slavapestov commented:
@theblixguy That looks like a legitimate error caught by your change. I'm taking a look now.

slavapestov commented:
Yeah, your change found a legitimate problem. Congratulations! :)

The following segfaults:

```
protocol A { 
  func foo()
}

protocol B: A {}

class Foo {
  func someMethod<T: A>(arg: T) {
    arg.foo() 
  }
}

class Bar: Foo {
  override func someMethod<T: B>(arg: T) {
    arg.foo()
  }
}

func f(_ foo: Foo) {
  foo.someMethod(arg: CA())
}

struct CA : A {
  func foo() {
    print("Hello world")
  }
}

f(Bar())
```

slavapestov commented:
@theblixguy This should unblock your PR: https://github.com/apple/swift/pull/25941

theblixguy committed:
[Typechecker] Bail out if we don't have any generic params in the GSB, otherwise we're gonna trip an assertion

theblixguy commented:
Thanks! Let’s run the CI again, hopefully the tests should pass now

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -945,6 +1040,30 @@ bool OverrideMatcher::checkOverride(ValueDecl *baseDecl,
+  // If the generic signatures are different, then complain.
+  auto newSig = getNewGenericSignature(baseDecl, decl);
+
+  if (newSig && derivedGenericCtx) {
```
Is it possible for derivedGenericCtx == nullptr && newSig != nullptr? Maybe check if derivedGenericCtx and baseGenericCtx are both non-null before calling getNewGenericSignature()

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -945,6 +1040,30 @@ bool OverrideMatcher::checkOverride(ValueDecl *baseDecl,
+
+    if (derivedSig) {
+      auto satOne = derivedSig->requirementsNotSatisfiedBy(newSig).empty();
+      auto satTwo = newSig->requirementsNotSatisfiedBy(derivedSig).empty();
```
it occurred to me that it might be better to do `newSig->getCanonicalSignature() == derivedSig->getCanonicalSignature()`

slavapestov replied:
Ignore this part, it only makes sense if we're doing an exact comparison here, but we really just want satOne. If satTwo is false, the code is still valid but you need to thunk.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+      }
+    }
+
+    if (builder.getGenericParams().empty()) {
```
This is easy to determine before creating the GSB -- if derivedClass->getGenericSignature() is null and !baseGenericCtx->isGeneric(), there won't be any generic parameters.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+  auto subMap = derivedClass->getSuperclass()->getContextSubstitutionMap(
+      derivedClass->getModuleContext(), baseClass);
+
+  if (baseGenericCtx->getGenericSignature() != nullptr) {
```
An early return is better than a level of indentation that goes until the end of the function

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+
+    if (auto derivedGenericCtx = derived->getAsGenericContext()) {
+      if (derivedGenericCtx->isGeneric()) {
+        for (auto param : *derivedGenericCtx->getGenericParams()) {
```
I think there's a subtle problem here. This should add the generic parameters of the base method, but translate them as with substFn.

Otherwise, you might get a crash or bad behavior if the override has a different number of generic parameters than the base method, since below you're adding requirements from the base signature.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+      depth = genericSig->getGenericParams().back()->getDepth() + 1;
+
+    GenericSignatureBuilder builder(ctx);
+    builder.addGenericSignature(derivedClass->getGenericSignature());
```
Shouldn't you be checking if derivedClass->getGenericSignature() is non-null above, instead of baseGenericCtx->getGenericSignature()?

slavapestov commented on include/swift/AST/GenericSignature.h:
```diff
@@ -272,8 +273,8 @@ class alignas(1 << TypeAlignInBits) GenericSignature final
-  SmallVector<Requirement, 4> requirementsNotSatisfiedBy(
-                                               GenericSignature *otherSig);
+  SmallVector<Requirement, 4>
+  requirementsNotSatisfiedBy(GenericSignature *otherSig);
```
Maybe we should back out this part

slavapestov commented on lib/AST/GenericSignature.cpp:
```diff
@@ -587,8 +587,8 @@ bool GenericSignature::isRequirementSatisfied(Requirement requirement) {
-SmallVector<Requirement, 4> GenericSignature::requirementsNotSatisfiedBy(
-                                                 GenericSignature *otherSig) {
+SmallVector<Requirement, 4>
+GenericSignature::requirementsNotSatisfiedBy(GenericSignature *otherSig) {
```
Ditto

slavapestov reviewed (approved):
I studied the code more closely and have even more suggestions, but if you want you can do it after merging this if the existing tests pass.

theblixguy commented:
Hmm source compat failed because of a bunch of failures.

slavapestov commented:
Can we just relax your check? We want each derived method requirement to appear in the base signature but not vice versa.

For now, we can allow but miscompile the case where they don’t match. To make it work properly, we’ll need to move getNewGenericSignature() to a common location, and in the code that checks if a method override needs a thunk, we will check the other direction, that the base method’s requirements are fully satisfied by the derived method’s signature. If they’re not, we need a thunk.

slavapestov commented:
(So the StdlibCollectionUnittest failure was invalid code; the derived method had more requirements than the base method. The SwiftNIO failure is the case we’d like to support.)

theblixguy commented:
> Can we just relax your check? We want each derived method requirement to appear in the base signature but not vice versa.

So we only need to do `derivedSig->requirementsNotSatisfiedBy(newSig).empty()`?

> in the code that checks if a method override needs a thunk, we will check the other direction, that the base method’s requirements are fully satisfied by the derived method’s signature. If they’re not, we need a thunk.

So we call `getNewGenericSignature()` again but this time, we do `baseSig->requirementsNotSatisfiedBy(newSig).empty()`? Is that going to be in `requiresNewVTableEntry()`?

slavapestov commented:
> So we only need to do newSig->getCanonicalSignature() == derivedSig->getCanonicalSignature() in checkOverride()?

No, that would check that both have identical requirements. You still want to use getRequirementsNotSatisfiedBy().

> So we call getNewGenericSignature() again but this time, we do newSig->getCanonicalSignature() == baseSig->getCanonicalSignature()? Is that going to be in requiresNewVTableEntry()?

Here you check if the base has any requirements not satisfied by the base using getRequirementsNotSatisfiedBy(), at the end after the call to matches().

We should also consider caching the result of getNewGenericSignature() in the ASTContext, as well as giving it a more descriptive name. Probably getOverrideGenericSignature() is OK.

The cache key is a bit tricky. A simple implementation would just use the derived method as the key. However, many methods will have the same signature and we're going to keep recomputing it unnecessarily. So a minimal key would contain the base method generic signature, the derived class's superclass type and the derived class generic signature. These three determine all the other inputs I think. You can derived the base *class* generic signature from the superclass type, then build your 'override' signature.

theblixguy committed:
[Typechecker] Change method name to getOverrideGenericSignature

theblixguy committed:
[Typechecker] Ensure both base and derived generic contexts are not null before calling getOverrideGenericSignature()

theblixguy committed:
[Typechecker] Only check for derivedSig->requirementsNotSatisfiedBy(newSig)

theblixguy committed:
[Typechecker] Check for empty generic parameters early

theblixguy committed:
[Typechecker] Use early return if the base generic context does not have a generic signature

theblixguy commented:
Okay, let's get this in first and then I will put up a follow-up PR to address your other suggestions and improvements :)

> Here you check if the base has any requirements not satisfied by the base using getRequirementsNotSatisfiedBy(), at the end after the call to matches()

Is that in `requiresNewVTableEntry`? There's a check for `!derivedInterfaceTy->matches` in the end.

theblixguy committed:
[Diagnostics] Fix a typo in diagnostic message

theblixguy committed:
[Test] Remove diagnostic from a test case

theblixguy committed:
[AST] Undo unnecessary indentation changes

theblixguy committed:
[AST] Emit vtable thunk if base generic requirements are not satisfied by derived

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -6486,6 +6486,12 @@ static bool requiresNewVTableEntry(const AbstractFunctionDecl *decl) {
   auto overrideInterfaceTy = selfInterfaceTy->adjustSuperclassMemberDeclType(
       base, decl, baseInterfaceTy);
 
+  if (!base->getGenericSignature()
```
Let's omit this part entirely and continue to miscompile this case for now.

When you do make this change, you actually want getOverrideGenericSignature() instead of base->getGenericSignature(), and we also need SILGen tests that verify a vtable thunk is emitted instead of a direct reference to the derived method.

theblixguy replied:
Okay, done!

theblixguy committed:
Revert "[AST] Emit vtable thunk if base generic requirements are not satisfied by derived"

This reverts commit e31a360ae961c663556d08694d5790ffb38e782f.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

theblixguy commented:
Interesting, still getting the same source compat [failure](https://ci.swift.org/job/swift-PR-source-compat-suite/3182//artifact/swift-source-compat-suite/FAIL_fluent_4.2_BuildSwiftPackage.log) as earlier:

slavapestov commented:
@theblixguy Maybe dump newSig in the failing case and see if it was built incorrectly.

slavapestov commented:
In fact, adding newSig to the diagnostic would be helpful because it is the expected signature of the override (the base method's signature is not directly usable in the generic context of the derived method).

theblixguy committed:
[Typechecker] Add the new generic signature to the diagnostic message

theblixguy commented:
@slavapestov I think I have managed to reduce the failure code to this example:

```swift
protocol FakeChannelOption {}

class FakeBaseSocket {}
class FakeSocket: FakeBaseSocket {}

class FakeBaseSocketChannel<T: FakeBaseSocket> {
  func fakeGetOption0<Option: FakeChannelOption>(_ option: Option) {}
}

class FakeSocketChannel: FakeBaseSocketChannel<FakeSocket> {
  override func fakeGetOption0<Option: FakeChannelOption>(_ option: Option) {}
}
```

where I get `overridden method 'fakeGetOption0' has generic signature <Option where Option : FakeChannelOption> which is incompatible with base method's generic signature <T, Option where T : FakeBaseSocket, Option : FakeChannelOption>; expected generic signature to be <Option>`.

(which is close to the original error - `overridden method 'getOption0' has generic signature <T where T : ChannelOption> which is incompatible with base method's generic signature <T, T where T : BaseSocket, T : ChannelOption>`)

slavapestov commented:
Notice how in the failing case, the base class is generic, while the derived class is not; while both the base method and derived method are generic. It looks like the base method's requirements are not being remapped correctly in this case, so they're being dropped.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+
+  if (auto derivedGenericCtx = derived->getAsGenericContext()) {
+    if (derivedGenericCtx->isGeneric()) {
+      for (auto param : *derivedGenericCtx->getGenericParams()) {
```
I think to fix that test failure, you need to iterate over the generic parameters of the base method, not the derived method here. See https://github.com/apple/swift/pull/24484#discussion_r299743599

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
+  }
+  unsigned depth = 0;
+
+  if (auto *genericSig = baseClass->getGenericSignature())
```
Should this be derivedClass->getGenericSignature()?

theblixguy replied:
Fixed!

slavapestov commented:
It looks like `superclassDepth == depth` is always the case, so `substFn` does not correctly remap generic arguments from the base class signature to the derived class signature. As a result `reqt.subst()` fails and we drop the requirements.

theblixguy committed:
[Typechecker] Add base method's generic parameters instead of derived

theblixguy commented:
@slavapestov Okay, so I have changed the code so it adds the base method's generic params instead of derived, although it causes the same assertion being hit as before (no generic params in GSB i.e. `Assertion failed: (params.empty() || ((Key.Depth == params.back()->getDepth() && Key.Index == params.back()->getIndex() + 1) || (Key.Depth > params.back()->getDepth() && Key.Index == 0)))`). Maybe I can add a check to bail out if there's no params?

theblixguy commented:
This might be because of the issue with `substFn` as you mentioned because we're dropping requirements. Is the issue with the comparison (gp->getDepth() == superclassDepth) or the way we're  calculating the depth (gp->getDepth() - superclassDepth + depth)?

slavapestov commented:
You need to add the derived method's generic params as before -- the problem is with the calculation of `depth`. Right now both `depth` and `superclassDepth` are computed from the base class's signature. However I think `depth` should be computed from the derived class's signature. For clarity you can rename `depth` to `derivedDepth` and `superclassDepth` to `baseDepth`.

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -970,18 +970,14 @@ static GenericSignature *getOverrideGenericSignature(ValueDecl *base,
   unsigned depth = 0;
 
-  if (auto *genericSig = baseClass->getGenericSignature())
+  if (auto *genericSig = derivedClass->getGenericSignature())
```
This is right

slavapestov commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -970,18 +970,14 @@ static GenericSignature *getOverrideGenericSignature(ValueDecl *base,
-        builder.addGenericParameter(param);
-      }
-    }
+  for (auto param : *baseGenericCtx->getGenericParams()) {
```
Let's revert this part though

theblixguy committed:
[Typechecker] Revert to using derivedGenericCtx params and rename depth and superclassDepth vaeiables

theblixguy commented:
> Right now both depth and superclassDepth are computed from the base class's signature

Hmm that doesn't appear to be the case as depth was calculated from `derivedClass->getGenericSignature()` and superclassDepth from `baseClassSig` (which was set to `baseClass->getGenericSignature()` earlier).

theblixguy committed:
[Typechecker] Fix an error due to use of old variable name

theblixguy commented:
@slavapestov Seems like the problem is fixed, I no longer get the error on the previous example, so I think the compatibility tests should pass now. Let's run the CI to verify 🙏

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

theblixguy commented:
@xedin @slavapestov could you rerun source compat tests again? The failure is unrelated (seems like two packages failed to build due to some errors `error: 'url(with:)' is only available in macOS 10.11 or newer` and `error: 'NSTextView' is incompatible with 'weak' references`)

xedin commented:
Looks like it just has to be XFailed first, I am going to do it later today.

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
@theblixguy My only remaining concern is that we're going to create more GenericSignatureBuilders than before. They're pretty heavy weight. If there's a performance regression here, we'll need to add caching in getOverrideGenericSignature(). Either way I'm happy merging this before we fix that, but there's no way to run performance tests on an already-merged PR.

theblixguy commented:
Yep, there’s a small regression: https://ci.swift.org/job/swift-PR-compiler-performance-macOS/430/artifact/comment.md

Let’s run source compat and once it passes, merge this PR and I’ll put up a follow up PR with caching!

slavapestov commented:
@theblixguy Yes, please take a look at the caching.

I think this fix is worth a changelog entry too.

theblixguy commented:
@slavapestov Okay, I'll put up a new PR shortly :)

Is it worth cherry picking this to 5.1 as well? Or is it too risky?
